### ISSUE TITLE
[iOS] ScrollPager and onIndexChange

### ISSUE THREAD

**Issue by solomonhawk** [react-native-community/react-native-tab-view#993]
### Current behaviour

With a simple slider like the example, adding the sample code from the readme to enable the `ScrollPager` on iOS results in `onIndexChange` never getting called. Is this intended (necessary for some reason that's not immediately obvious)?

### Expected behaviour

I expect `onIndexChange` to still get called when passing `renderPager={props => <ScrollPager {...props} />}`.

### Code sample

Here's an example snack that demonstrates the issue. **Make sure you are viewing this on the iOS device emulator**. Try uncommenting line 31 (`// renderPager={props => <ScrollPager {...props} />}`) in `App.js`.

Swiping between the tabs results in an alert with the index (`onIndexChange` fires `alert`). After uncommenting the line to enable usage of the `ScrollPager`, the alert no longer fires.

https://snack.expo.io/HkQl8ETI8

### What have you tried

I looked around to see whether there was another way I could observe the ending index after a swipe but didn't find a good way to do so. I can pass `onSwipeEnd` to `ScrollPager` but it gets called with no parameters and I'm left only with access to the `Animated.Value` `position` which it seems like I can't just synchronously read on the JS thread (am I wrong on that)?

#### Possible Solutions

1. Don't use `ScrollPager` - this would be a bummer because it does give a bit of a nicer interaction on iOS
2. ??? Taking a look at which animated node pipeline processing logic we might be able to pull in from `Pager` to observe `position` and `call` back out.

### Your Environment

| software                     | version
| ---------------------------- | -------
| ios or android               | ios 11
| react-native                 | 0.61.5
| react-native-tab-view        | 2.13
| react-native-gesture-handler | 1.6.1
| react-native-reanimated      | 1.7.1
| node                         | 12.13.0
| npm or yarn                  | yarn 1.21.1

**Comment by satya164** [comment]
cc @osdnk

**Comment by solomonhawk** [comment]
I'm guessing, naively, that part of this is due to the fact that when using `ScrollPager`, the x pan position is "owned" by the `ScrollView` and we synchronize an animated value with its `contentOffset.x` during `onScroll`. This should still provide an opportunity to `call` back out when `position` comes to a rest.

**Comment by solomonhawk** [comment]
I poked around a bit and found something that patches the issue but I'm not familiar enough with things to know whether it is correct.

```js
<Animated.Code
  exec={onChange(
    this.position,
    cond(eq(modulo(this.position, this.props.layout.width), 0), [
      call([this.relativePosition], ([relativePosition]) =>
        onIndexChange(relativePosition)
      ),
    ])
  )}
/>
```

Diff:

```diff
diff --git a/src/ScrollPager.tsx b/src/ScrollPager.tsx
index 664a586..037cce5 100644
--- a/src/ScrollPager.tsx
+++ b/src/ScrollPager.tsx
@@ -4,7 +4,7 @@ import Animated from 'react-native-reanimated';
 import { Props } from './Pager';
 import { Route, Listener } from './types';
 
-const { event, divide, Value } = Animated;
+const { event, divide, onChange, cond, eq, modulo, call, Value } = Animated;
 
 type State = {
   initialOffset: { x: number; y: number };
@@ -136,6 +136,7 @@ export default class ScrollPager<T extends Route> extends React.Component<
       onSwipeStart,
       onSwipeEnd,
       overscroll,
+      onIndexChange,
       navigationState,
     } = this.props;
 
@@ -157,38 +158,51 @@ export default class ScrollPager<T extends Route> extends React.Component<
       removeListener: this.removeListener,
       jumpTo: this.jumpTo,
       render: children => (
-        <Animated.ScrollView
-          pagingEnabled
-          directionalLockEnabled
-          keyboardDismissMode="on-drag"
-          keyboardShouldPersistTaps="always"
-          overScrollMode="never"
-          scrollToOverflowEnabled
-          scrollEnabled={this.props.swipeEnabled}
-          automaticallyAdjustContentInsets={false}
-          bounces={overscroll}
-          scrollsToTop={false}
-          showsHorizontalScrollIndicator={false}
-          scrollEventThrottle={1}
-          onScroll={this.onScroll}
-          onScrollBeginDrag={handleSwipeStart}
-          onScrollEndDrag={handleSwipeEnd}
-          onMomentumScrollEnd={this.onScroll}
-          contentOffset={this.initialOffset}
-          style={styles.container}
-          contentContainerStyle={
-            layout.width
-              ? {
-                  flexDirection: 'row',
-                  width: layout.width * navigationState.routes.length,
-                  flex: 1,
-                }
-              : null
-          }
-          ref={this.scrollViewRef}
-        >
-          {children}
-        </Animated.ScrollView>
+        <>
+          <Animated.ScrollView
+            pagingEnabled
+            directionalLockEnabled
+            keyboardDismissMode="on-drag"
+            keyboardShouldPersistTaps="always"
+            overScrollMode="never"
+            scrollToOverflowEnabled
+            scrollEnabled={this.props.swipeEnabled}
+            automaticallyAdjustContentInsets={false}
+            bounces={overscroll}
+            scrollsToTop={false}
+            showsHorizontalScrollIndicator={false}
+            scrollEventThrottle={1}
+            onScroll={this.onScroll}
+            onScrollBeginDrag={handleSwipeStart}
+            onScrollEndDrag={handleSwipeEnd}
+            onMomentumScrollEnd={this.onScroll}
+            contentOffset={this.initialOffset}
+            style={styles.container}
+            contentContainerStyle={
+              layout.width
+                ? {
+                    flexDirection: 'row',
+                    width: layout.width * navigationState.routes.length,
+                    flex: 1,
+                  }
+                : null
+            }
+            ref={this.scrollViewRef}
+          >
+            {children}
+          </Animated.ScrollView>
+
+          <Animated.Code
+            exec={onChange(
+              this.position,
+              cond(eq(modulo(this.position, this.props.layout.width), 0), [
+                call([this.relativePosition], ([relativePosition]) =>
+                  onIndexChange(relativePosition)
+                ),
+              ])
+            )}
+          />
+        </>
       ),
     });
   }
```

**Comment by orofbrown** [comment]
I'm also experiencing this issue where passing in an arg for `renderPager` isn't doing anything

**Comment by KoboBunny** [comment]
Hey @solomonhawk 

I'm trying to apply a patch using patch-package where I edit the TSX files in /src as opposed to the JS files in module /lib/module

I noticed you also applied your patch to the TSX files and I was curious how you were able to get it to work?  When I create my patch, the files in /lib/module don't change and the changes are also not reflected in my app.  

Am I missing a build or compile step to get react-native-tab-view to generate new JS files from the TSX files I created my patch for?

Any help would be greatly appreciated!
